src/dataflow/src/render/join/delta_join.rs

```diff
@@ -530,27 +480,38 @@ use differential_dataflow::trace::Cursor;
 use differential_dataflow::trace::TraceReader;
 use differential_dataflow::Collection;
 
-/// Constructs a `lookup_map` from supplied arguments.
+/// Constructs a `half_join` from supplied arguments.
 ///
 /// This method exists to factor common logic from four code paths that are generic over the type of trace.
-fn build_lookup<G, Tr>(
-    updates: Collection<G, Row>,
+/// The `comparison` function should either be `le` or `lt` depending on which relation comes first in the
+/// total order on relations (in order to break ties consistently).
+///
+/// The input and output streams are of pairs `(data, time)` where the `time` component can be greater than
+/// the time of the update. This operator may manipulate `time` as part of this pair, but will not manipulate
+/// the time of the update. This is crucial for correctness, as the total order on times of updates is used
+/// to ensure that any two updates are matched at most once.
+fn build_halfjoin<G, Tr, CF>(
+    updates: Collection<G, (Row, G::Timestamp)>,
     trace: Arranged<G, Tr>,
     prev_key: Vec<MirScalarExpr>,
+    comparison: CF,
```

Comment:
this is a super minor thought but if the comparison function can only be one of two things maybe this arg should be a bool / enum instead of a closure?

Reply:
It could, but then the runtime behavior would involve another conditional check. This approach results in two compiled instances with specialized logic, which we choose between at dataflow construction time.